### DISCUSSION TITLE
multithreading with PEGTL

### DISCUSSION THREAD

**Discussion by Shaheen47** [taocpp/PEGTL#176] [Q&A] [answered]
Greetings,
I am implementing a stream parser and having some producer-consumer problem.
I have two threads, one main thread and one thread that runs the parse function. 
there is a concurrent queue and I am adding elements to this queue inside an action(which runs by the parsing thread).
the main thread has some functions for getting the values from the concurrent queue.
this queue has a capacity. and what I would like to do is to stop the parsing thread when the number of elements in the concurrent queue reaches a threshold and then continue it when the main thread finds that the number of elements in the concurrent queue is less than another threshold.
what in my mind is to use condition variables for waiting inside the action in the parsing thread. and notify on the main thread.
are there any risks of doing this inside the actions ? for example does this conflicts with any internal PEGTl functionality that could produce undefined behavior when I implement my synchronization method? Do you recommend another approach here?
Thanks very much in advance

**Comment by ColinH** [comment, marked as answer]
The PEGTL is completely thread agnostic, it shouldn't care about whatever it is that you're doing as long as you are respecting all the basic rules of multi-threaded programming.

If your concurrent queue blocks the producer on a condition variable when you try to push into a queue that is already full, and the consumer signals that variable when it takes something out of the queue/a full queue, you should get exactly the behaviour that you need without any additional effort.
